rpc listtransactions new argument options (paginatebypointer impl)

hosseinzoda opened the pull request:
Added fifth param to `listtransactions` named `options`, `options` may be an object containing `paginatebypointer` (boolean default: false) and `nextpagepointer` (string default: OMITTED).

With `paginatebypointer` output will have the following changes.

1. Return transactions is ordered by most recent transactions. Though the default does reverse the order after transactions are fetched and clipped.
2. `skip` argument has no effect. Instead `nextpagepointer` will be used for pagination.
3. Return value is an object containing, records (array of txs) and nextpagepointer (string)

andrewtoth commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1500,6 +1500,267 @@ UniValue listtransactions(const JSONRPCRequest& request)
+            "[\n"
+            "  {\n"
+            "    \"address\":\"address\",    (string) The bitcoin address of the transaction.\n"
+            "    \"category\":\"send|receive\", (string) The transaction category.\n"
```
There are more categories than just `send` and `receive`. See https://github.com/bitcoin/bitcoin/pull/14653

hosseinamin replied:
I was using branch 0.17 , and when i upgraded to master. help file was from the old one. will fix

hosseinamin replied:
I see original `lsittransactions` only contains `send|receive`. @andrewtoth Please inform me. What are the other values. I guess there's `internal`.

MeshCollider replied:
The other 3 are generate, immature and orphan.

hosseinamin replied:
okay. I have made the change. i will add it after review.

hosseinamin commented:
@promag `listtransactions` has been used in many tests. I don't know what will happen. If change number one gets added to it.

> 1. Return transactions is ordered by most recent transactions. Though the original rpc does reverse the order after transactions are fetched and clipped.

There's an odd reverse call at `listtransactions` which is not in this one.
https://github.com/bitcoin/bitcoin/blob/fc1710fbf3ce840f0b647913be8b821089b00f5e/src/wallet/rpcwallet.cpp#L1494

Other than that adding `nextpagepointer` will break compatibility with `listtransactions` params. Only if one accepts it as interger and string and keeps the behaviour the same.

jonasschnelli commented:
I general I like the idea of a pointer (seems more reliable) and the sort.
Not sure about cloning the call.

kristapsk commented:
> @promag `listtransactions` has been used in many tests. I don't know what will happen. If change number one gets added to it.

I would not worry about tests, you should anyway run the functional tests yourself if you change something in RPC. Sort order could be added as an additional optional boolean parameter IMO. As well as other changes.

hosseinamin commented:
Who thinks this line should get removed from `listtransactions`?

https://github.com/bitcoin/bitcoin/blob/fc1710fbf3ce840f0b647913be8b821089b00f5e/src/wallet/rpcwallet.cpp#L1494

sipa commented:
@hosseinamin Well we can't remove that line; there are likely end users who rely on its behavior, even if it is suboptimal. We can make it optional by adding an extra flag, though.

kristapsk commented:
Breaking userspace is bad, so changing default behaviour of any RPC call should be avoided.

hosseinamin commented:
That's why i did cloned the rpc call. This is how new functionality can come in without messing with userspace.

jonasschnelli commented:
> That's why i did cloned the rpc call. This is how new functionality can come in without messing with userspace.

Cloning is one way, backward compatible parameterising would be another (see how the account system was removed).

hosseinamin commented:
Okay. How's this.

1. `skip` param can morph to `nextpagepointer` when string is given.
2. adding `nonlegacy` parameter at the end to avoid messing with userspace.

hosseinamin commented:
I think `skip` accepts string as input. It will convert to integer. A bit more work is needed to detect between the two.

hosseinamin commented:
@jonasschnelli Removed listransactions2, And did parametrize it.
I will try later to write some tests for it.

practicalswift commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1444,16 +1478,122 @@ UniValue listtransactions(const JSONRPCRequest& request)
+                } else if (i == 1) {
+                    try {
+                        nextVOut = std::stoul(s);
+                    } catch (std::invalid_argument &e) {
```
Drop `&e`?

practicalswift commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1444,16 +1478,122 @@ UniValue listtransactions(const JSONRPCRequest& request)
+                        nextVOut = std::stoul(s);
+                    } catch (std::invalid_argument &e) {
+                        break;
+                    } catch (std::out_of_range &e) {
```
Same here?

kristapsk commented:
I'm not sure a single parameter that changes multiple things is a right way to go. Wouln't be better to be able to specify sorting order and paging behaviour independently? Besides I don't like "old behaviour vs new behaviour" boolean also because in future there could be a different one "new new behaviour".

hosseinamin commented:
Okay. I agree. Want to change `newversion` to `flags`. Though don't have time to do it now.

practicalswift commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1294,8 +1294,13 @@ static void MaybePushAddress(UniValue & entry, const CTxDestination &dest)
  * @param  ret            The UniValue into which the result is stored.
  * @param  filter_ismine  The "is mine" filter flags.
  * @param  filter_label   Optional label string to filter incoming transactions.
+ * @param  afterVOut      start adding after afterVOut
```
`afterVOut` is not part of the declaration?

hosseinamin replied:
Should get removed. Left out from some dead coding.

practicalswift commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1444,16 +1478,122 @@ UniValue listtransactions(const JSONRPCRequest& request)
+                // skip processing until txid match found
+                while (it != txOrdered.rend()) {
+                    CWalletTx *const pwtx = (*it++).second;
+                    if (nextHash == pwtx->GetHash()) {
```
Nit: Reduce indentation by reversing logic. `continue` if `nextHash != pwtx->GetHash() instead`.

laanwj commented:
@hosseinamin Can you please update the PR title and opening post to cover the current changes?

hosseinamin commented:
> @hosseinamin Can you please update the PR title and opening post to cover the current changes?

Where can i open a post? (i did edit the current description)

laanwj commented:
> @laanwj Where can i open a post? (i did edit the current description)

Thanks. That's what I meant, with "opening post" I mean the first post.

laanwj commented:
Travis error:
```
The locale dependent function stoul(...) appears to be used:
src/wallet/rpcwallet.cpp:                        nextVOut = std::stoul(s);
src/wallet/rpcwallet.cpp:                        nextIsSend = std::stoul(s) == 1;
```
PLease use our own functions `ParseInt32` or `ParseInt64` instead.

laanwj commented:
Linter error from travis:
```
This diff appears to have added new lines with trailing whitespace.
The following changes were suspected:
diff --git a/src/wallet/rpcwallet.cpp b/src/wallet/rpcwallet.cpp
@@ -1455,2 +1492,113 @@ UniValue listtransactions(const JSONRPCRequest& request)
+
+
+
+
+
^---- failure generated from test/lint/lint-whitespace.sh
```

hosseinamin commented:
This PR needs review.

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1403,10 +1434,15 @@ UniValue listtransactions(const JSONRPCRequest& request)
-                    {"skip", RPCArg::Type::NUM, /* opt */ true, /* default_val */ "0", "The number of transactions to skip"},
+                    {"skip", RPCArg::Type::NUM, /* opt */ true, /* default_val */ "0", "The number of transactions to skip, In case of flags contain nextpagepointer, fThe value of it is a string given by listtransactions."},
                     {"include_watchonly", RPCArg::Type::BOOL, /* opt */ true, /* default_val */ "false", "Include transactions to watch-only addresses (see 'importaddress')"},
+                    {"flags", RPCArg::Type::STR, /* opt */ true, /* default_val */ "", "list of flags separated by | which they are [ nextpagepointer ]"},
```
nit: I think we either use a json object called `options` or a named argument directly. A specially serialized `flags` seems inappropriate given that we have all json types and named arguments available.

Commit:
rpc listtransactions new argument options (paginatebypointer impl)

Added fifth param to `listtransactions` named `options`, `options` may be an object containing `paginatebypointer` (boolean default: false) and `nextpagepointer` (string default: OMITTED).

With `paginatebypointer` output will have the following changes.

1. Return transactions is ordered by most recent transactions. Though the default does reverse the order after transactions are fetched and clipped.
2. `skip` argument has no effect. Instead `nextpagepointer` will be used for pagination.
3. Return value is an object containing, records (array of txs) and nextpagepointer (string)

hosseinamin commented:
Hey guys. This commit is ready for merge.

practicalswift commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1400,10 +1431,20 @@ UniValue listtransactions(const JSONRPCRequest& request)
             "              with the specified label, or \"*\" to disable filtering and return all transactions."},
                     {"count", RPCArg::Type::NUM, /* default */ "10", "The number of transactions to return"},
-                    {"skip", RPCArg::Type::NUM, /* default */ "0", "The number of transactions to skip"},
+                    {"skip", RPCArg::Type::NUM, /* default */ "0", "The number of transactions to skip, If \"usesnextpagepointer\" option is true, skip wont have an effect."},
```
"Wont" whould be "won't" :-)

adamjonas commented:
Checking in on this. It's been a while since there was a rebase and there doesn't appear to be strong conviction from the reviewers on the concept/approach.

hosseinamin commented:
I wasted my time on this pull-request. No need for the merge anymore. Lets dump it to garbage.

kristapsk commented:
I still think this would have been a good functionality conceptually.
